nodes/core/storage/50-file.js

```diff
@@ -130,25 +137,75 @@ module.exports = function(RED) {
                             var packet = node.data.shift()
                             node.wstream.write(packet.data, function() {
                                 node.send(packet.msg);
+                                done();
                             });
 
                         } else {
                             // Dynamic filename - write and close the stream
                             var packet = node.data.shift()
                             node.wstream.end(packet.data, function() {
                                 node.send(packet.msg);
+                                delete node.wstream;
+                                delete node.wstreamIno;
+                                done();
                             });
-                            delete node.wstream;
-                            delete node.wstreamIno;
                         }
                     }
                 }
             }
+        }
+
+        function processQ(queue) {
+            var msg = queue[0];
+            processMsg(msg, function() {
+                queue.shift();
+                if (queue.length > 0) {
+                    processQ(queue);
+                }
+            });
+        }
+
+        this.on("input", function(msg) {
+            var msgQueue = node.msgQueue;
+            if (msgQueue.push(msg) > 1) {
+                // pending write exists
+                return;
+            }
+            processQ(msgQueue);
+            if (node.closing) {
+                closeNode();
+            }
         });
-        this.on('close', function() {
+
+        function closeNode() {
             if (node.wstream) { node.wstream.end(); }
             if (node.tout) { clearTimeout(node.tout); }
             node.status({});
+            var callbacks = node.closeCallbacks;
+            node.closeCallbacks = [];
+            node.closing = false;
+            for (cb in callbacks) {
+                cb();
+            }
+        }
+
+        this.on('close', function(arg1, arg2) {
+            var cb  = arg2 ? arg2 : arg1;
+            if (cb) {
+                node.closeCallbacks.push(done);
```

Comment:
A node can only be closed once - there will only be one callback. So rather than `node.closeCallbacks` as an array, I suggest `cb` get stored as `node.closeCallback`.